File: lib/java/protocol/src/main/java/org/thryft/protocol/AbstractOutputProtocol.java
/*******************************************************************************
 * Copyright (c) 2013, Minor Gordon
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 ******************************************************************************/

package org.thryft.protocol;

import java.math.BigDecimal;
import java.util.Date;

import org.apache.commons.codec.binary.Base64;
import org.thryft.Base;
import org.thryft.native_.EmailAddress;
import org.thryft.native_.Uri;
import org.thryft.native_.Url;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.google.common.primitives.UnsignedInteger;
import com.google.common.primitives.UnsignedLong;

public abstract class AbstractOutputProtocol implements OutputProtocol {
    @Override
    public void flush() throws OutputProtocolException {
    }

    @Override
    public void writeBinary(final byte[] value) throws OutputProtocolException {
        writeString(Base64.encodeBase64String(value));
    }

    @Override
    public void writeDateTime(final Date value) throws OutputProtocolException {
        writeI64(value.getTime());
    }

    @Override
    public void writeDecimal(final BigDecimal value)
            throws OutputProtocolException {
        writeString(value.toString());
    }

    @Override
    public void writeEmailAddress(final EmailAddress value)
            throws OutputProtocolException {
        writeString(value.toString());
    }

    @Override
    public void writeEnum(final Enum<?> value) throws OutputProtocolException {
        writeString(value.toString().toUpperCase());
    }

    @Override
    public void writeFieldBegin(final String name, final Type type,
            final short id) throws OutputProtocolException {
        throw new UnsupportedOperationException();
    }

    @Override
    public void writeFieldEnd() throws OutputProtocolException {
        throw new UnsupportedOperationException();
    }

    @Override
    public void writeFieldStop() throws OutputProtocolException {
        throw new UnsupportedOperationException();
    }

    @Override
    public void writeMessageBegin(final String name, final MessageType type,
            final int sequenceId) throws OutputProtocolException {
        writeStructBegin(name);
    }

    @Override
    public void writeMessageEnd() throws OutputProtocolException {
        writeStructEnd();
    }

    @Override
    public void writeMixed(final Object value) throws OutputProtocolException {
        if (value == null) {
            writeNull();
        } else if (value instanceof Boolean) {
            writeBool((Boolean) value);
        } else if (value instanceof Byte) {
            writeByte((Byte) value);
        } else if (value instanceof Date) {
            writeDateTime((Date) value);
        } else if (value instanceof BigDecimal) {
            writeDecimal((BigDecimal) value);
        } else if (value instanceof Double) {
            writeDouble((Double) value);
        } else if (value instanceof ImmutableList) {
            @SuppressWarnings("unchecked")
            final ImmutableList<Object> set = (ImmutableList<Object>) value;
            writeListBegin(Type.VOID, set.size());
            for (final Object element : set) {
                writeMixed(element);
            }
            writeListEnd();
        } else if (value instanceof ImmutableMap) {
            @SuppressWarnings("unchecked")
            final ImmutableMap<Object, Object> map = (ImmutableMap<Object, Object>) value;
            writeMapBegin(Type.VOID, Type.VOID, map.size());
            for (final ImmutableMap.Entry<Object, Object> entry : map
                    .entrySet()) {
                writeMixed(entry.getKey());
                writeMixed(entry.getValue());
            }
            writeMapEnd();
        } else if (value instanceof ImmutableSet) {
            @SuppressWarnings("unchecked")
            final ImmutableSet<Object> set = (ImmutableSet<Object>) value;
            writeSetBegin(Type.VOID, set.size());
            for (final Object element : set) {
                writeMixed(element);
            }
            writeSetEnd();
        } else if (value instanceof EmailAddress) {
            writeEmailAddress((EmailAddress) value);
        } else if (value instanceof Short) {
            writeI16((Short) value);
        } else if (value instanceof Integer) {
            writeI32((Integer) value);
        } else if (value instanceof Long) {
            writeI64((Long) value);
        } else if (value instanceof String) {
            writeString((String) value);
        } else if (value instanceof Url) {
            writeUrl((Url) value);
        } else if (value instanceof Base<?>) {
            ((Base<?>) value).write(this);
        } else {
            throw new UnsupportedOperationException(value.getClass()
                    .getCanonicalName());
        }
    }

    @Override
    public void writeSetBegin(final Type elementType, final int size)
            throws OutputProtocolException {
        writeListBegin(elementType, size);
    }

    @Override
    public void writeSetEnd() throws OutputProtocolException {
        writeListEnd();
    }

    @Override
    public void writeU32(final UnsignedInteger value)
            throws OutputProtocolException {
        writeI32(value.intValue());
    }

    @Override
    public void writeU64(final UnsignedLong value)
            throws OutputProtocolException {
        writeI64(value.longValue());
    }

    @Override
    public void writeUri(final Uri value) throws OutputProtocolException {
        writeString(value.toString());
    }

    @Override
    public void writeUrl(final Url value) throws OutputProtocolException {
        writeString(value.toString());
    }
}

File: lib/py/src/thryft/protocol/_abstract_output_protocol.py
#------------------------------------------------------------------------------
# Copyright (c) 2013, Minor Gordon
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in
#       the documentation and/or other materials provided with the
#       distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
# CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
# OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.
#------------------------------------------------------------------------------


from time import mktime
from thryft.protocol._output_protocol import _OutputProtocol


class _AbstractOutputProtocol(_OutputProtocol):
    def writeByte(self, byte):
        self.writeI16(byte)
        return self

    def writeDateTime(self, date_time):
        self.write_i64(long(mktime(date_time.timetuple())) * 1000l)
        return self

    def writeDecimal(self, decimal):
        self.write_string(str(decimal))
        return self

    def writeEmailAddress(self, email_address):
        self.write_string(email_address)
        return self

    def writeI16(self, i16):
        self.write_i32(i16)
        return self

    def writeMixed(self, object_):
        if object_ is None:
            self.write_null()
        elif isinstance(object_, dict):
            self.write_map_begin(len(object_))
            for key, value in object_.iteritems():
                self.writeMixed(key)
                self.writeMixed(value)
            self.write_map_end()
        elif isinstance(object_, float):
            self.writeDouble(object_)
        elif isinstance(object_, frozenset):
            self.write_set_begin(len(object_))
            for item in object_:
                self.writeMixed(item)
            self.write_set_end()
        elif isinstance(object_, int):
            self.write_i32(object_)
        elif isinstance(object_, (list, tuple)):
            self.write_list_begin(len(object_))
            for item in object_:
                self.writeMixed(item)
            self.write_list_end()
        elif isinstance(object_, long):
            self.write_i64(object_)
        elif isinstance(object_, basestring):
            self.write_string(object_)
        elif hasattr(object_, 'write'):
            object_.write(self)
        else:
            raise TypeError(type(object_))
        return self

    def write_set_begin(self, *args, **kwds):
        self.write_list_begin()
        return self

    def write_set_end(self):
        self.write_list_end()
        return self

    def writeUrl(self, url):
        self.write_string(url)
        return self
